### PR TITLE
Adding Inverse Op

### PR DESCRIPTION
**Pull request by neginraoof**
Adding inverse op.
Similar to torch.inverse and numpy.linalg.inv(a), this op only supports square shape tensors, or matrices of shape [*, M, M].

### CODE REVIEW

**Comment by neginraoof** [comment]
@gramalingam 
cc @houseroad @postrational
Could you please help review this PR?
Thanks

File: docs/Operators.md [hunk 1]
```diff
@@ -6946,6 +6947,79 @@ expect(node, inputs=[x, s, bias], outputs=[y],
...
+
+<dl>
+<dt><tt>Y</tt> : T</dt>
+<dd>Output tensor</dd>
+</dl>
+
+#### Type Constraints
+
+<dl>
+<dt><tt>T</tt> : tensor(uint8), tensor(uint16), tensor(uint32), tensor(uint64), tensor(int8), tensor(int16), tensor(int32), tensor(int64), tensor(float16), tensor(float), tensor(double)</dt>
```
Comment (linkerzhang) [inline docs/Operators.md]: for 8-bit integers, do you mean quantized integers or just regular 8 bit integer? if it's quantized, same scale/offset used for input and output?

Input and output can be always the same type?

Comment (neginraoof) [reply docs/Operators.md]: Thanks! This PR is not targeting quantized tensors. 
I added the type constraint for output (same type as input) in the spec.

File: docs/Operators.md [hunk 2]
```diff
@@ -6946,6 +6947,79 @@ expect(node, inputs=[x, s, bias], outputs=[y],
 </details>
 
 
+### <a name="Inverse"></a><a name="inverse">**Inverse**</a>
+
+  Calculates inverse of a square matrix or batches of square matrices.
+  Inverse takes one input tensor of shape `[*, M, M]`, where `*` is zero or more batch dimensions,
+  and the inner-most 2 dimensions form square matrices.
+  The output is a tensor of shape `[*, M, M]`, containing the individual inverses of all input submatrices.
```
Comment (spandantiwari) [inline docs/Operators.md]: What happens when the matrix is rank-deficient, or more generally, ill-conditioned with a large condition number? We should clarify what the expectation is in that case.
I would recommend that we keep it flexible for backends and treat this as an undefined behavior. Any backend can choose detect the condition and throw exception, or simply return garbage result (possibly with a warning).

Comment (neginraoof) [reply docs/Operators.md]: Thanks for pointing this out. I agree that we should expect an exception from backend. Should I mention that rank-deficient matrices are not supported?

Comment (neginraoof) [reply docs/Operators.md]: I updated the spec noting that the op only supports fill-rank tensors.

File: onnx/defs/math/defs.cc [hunk 3]
```diff
@@ -2038,4 +2038,57 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+                (mat_w.dim_value() != mat_h.dim_value())) {
+              fail_shape_inference(
+                  "The inner-most 2 dimensions must have the same size (mat_w:",
+                  mat_w.dim_value(),
+                  " != mat_h:",
+                  mat_h.dim_value(),
+                  ").");
+            }
+
+            for (int i=0; i < rank ; ++i) {
```
Comment (gramalingam) [inline onnx/defs/math/defs.cc]: Can replace this loop by "propagateShapeFromInputToOutput(ctx, 0, 0);".

Comment (neginraoof) [reply onnx/defs/math/defs.cc]: Thanks. Replaced it.

**Comment by neginraoof** [comment]
cc @gramalingam @linkerzhang could you please help merge this PR?
Thanks

**Pull request merged**